File: day1/excercise 1/c#/Product.cs
public class Product
{
    private double _stock;

    public Product(string productName, double initialStock = 0)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(productName);
        ArgumentOutOfRangeException.ThrowIfNegative(initialStock);

        ProductName = productName;
        _stock = initialStock;
    }

    public string ProductName { get; }
    public double Stock => _stock;

    public void AddStock(double quantity)
    {
        if (quantity <= 0)
            throw new ArgumentOutOfRangeException(nameof(quantity));

        _stock += quantity;
    }

    public void Sell(double quantity)
    {
        if (quantity <= 0)
            throw new ArgumentOutOfRangeException(nameof(quantity));

        if (quantity > _stock)
            throw new InvalidOperationException("Insufficient stock.");

        _stock -= quantity;
    }
}

File: day1/excercise 1/node-express/product.js
class Product {
    #stock;

    constructor(productName, initialStock = 0) {
        if (!productName?.trim()) {
            throw new Error("Product name cannot be empty.");
        }

        if (initialStock < 0) {
            throw new Error("Initial stock cannot be negative.");
        }

        this.productName = productName;
        this.#stock = initialStock;
    }

    get stock() {
        return this.#stock;
    }

    addStock(quantity) {
        if (quantity <= 0) {
            throw new Error("Quantity must be greater than zero.");
        }

        this.#stock += quantity;
    }

    sell(quantity) {
        if (quantity <= 0) {
            throw new Error("Quantity must be greater than zero.");
        }

        if (quantity > this.#stock) {
            throw new Error("Insufficient stock.");
        }

        this.#stock -= quantity;
    }
}